mmtk/src/julia_scanning.rs

```diff
@@ -30,285 +36,359 @@ pub unsafe fn mmtk_jl_typeof(addr: Address) -> Address {
+                    let slot = obj.shift::<Address>(obj8_begin_loaded as isize);
+                    process_edge(closure, slot);
+                    obj8_begin = obj8_begin.shift::<u8>(1);
                 }
-            } else if (*layout).fielddesc_type() == 1 {
-                let layout_fields =
-                    Address::from_ptr(layout) + std::mem::size_of::<mmtk_jl_datatype_layout_t>();
-                let fielddesc_size = 2 << (*layout).fielddesc_type();
-                let obj16_start_addr = layout_fields + fielddesc_size * (*layout).nfields as usize;
-                let obj16_end_addr = obj16_start_addr + (npointers as usize * 2);
-                for elem_usize in
-                    (obj16_start_addr.as_usize()..obj16_end_addr.as_usize()).step_by(2)
-                {
-                    let elem = addr.to_mut_ptr::<*mut libc::c_uchar>();
-                    let index = Address::from_usize(elem_usize).load::<u16>();
-                    let slot: *mut libc::c_uchar = &mut *elem.offset(index as isize)
-                        as *mut *mut libc::c_uchar
-                        as *mut libc::c_uchar;
-                    process_edge(closure, Address::from_mut_ptr(slot));
+            } else if (*layout).fielddesc_type_custom() == 1 {
+                let mut obj16_begin = mmtk_jl_dt_layout_ptrs(layout);
+                let obj16_end = obj16_begin.shift::<u16>(npointers as isize);
+
+                while obj16_begin < obj16_end {
+                    let obj16_begin_loaded = obj16_begin.load::<u16>();
+                    let slot = obj.shift::<Address>(obj16_begin_loaded as isize);
+                    process_edge(closure, slot);
+                    obj16_begin = obj16_begin.shift::<u16>(1);
+                }
+            } else if (*layout).fielddesc_type_custom() == 2 {
+                let mut obj32_begin = mmtk_jl_dt_layout_ptrs(layout);
+                let obj32_end = obj32_begin.shift::<u32>(npointers as isize);
+
+                while obj32_begin < obj32_end {
+                    let obj32_begin_loaded = obj32_begin.load::<u32>();
+                    let slot = obj.shift::<Address>(obj32_begin_loaded as isize);
+                    process_edge(closure, slot);
+                    obj32_begin = obj32_begin.shift::<u32>(1);
                 }
-            } else if (*layout).fielddesc_type() == 2 {
-                unimplemented!();
             } else {
-                assert_eq!((*layout).fielddesc_type(), 3);
+                debug_assert!((*layout).fielddesc_type_custom() == 3);
+                unimplemented!();
             }
         }
     }
 }
 
+unsafe fn mmtk_scan_gcstack<EV: EdgeVisitor<JuliaVMEdge>>(
+    ta: *const mmtk_jl_task_t,
+    closure: &mut EV,
+) {
+    let stkbuf = (*ta).stkbuf;
+    let copy_stack = (*ta).copy_stack_custom();
+    // FIXME: the code below is executed COPY_STACKS has been defined in the C Julia implementation - it is on by default
```

Comment:
What does this comment mean?

Reply:
The C code has a `#ifdef COPY_STACKS` but I'm not sure how to do this in Rust (https://github.com/mmtk/julia/blob/f690aa3a5621bfa1d6a07f911818f203d3f8d650/src/gc.c#L2118).

It is set here in `options.h`: https://github.com/mmtk/julia/blob/f690aa3a5621bfa1d6a07f911818f203d3f8d650/src/options.h#L102

Reply:
Rust has `cfg` for conditional compilation. If necessary, you can turn the code into a block that is guarded with `#[cfg(feature = "julia_copy_stack")]`. As we build the Rust project separately from the C project, there is no way to get the flag from C in compile time. The only thing you can do is to have an assertion to make sure that `COPY_STACK` is defined in C.

Reply:
I'm not sure where to put the check. I've added [this function](https://github.com/udesou/mmtk-julia/blob/30f1e676b9e3d4768f99355f7368a28daee6269e/mmtk/src/util.rs#L100)  and I'm calling it before initialising MMTk (https://github.com/udesou/julia/blob/ed30d1c925865dd5c4ca2482701e20d6bc4700ec/src/mmtk-gc.c#L365). Would you suggest I do something different?